### PR TITLE
Iondotnet tree collapse

### PR DESCRIPTION
**Pull request by R-maan**

### CODE REVIEW

File: IonDotnet/Tree/IIonContainer.cs [hunk 1]
```diff
@@ -1,7 +1,16 @@
...
+    public interface IIonContainer
     {
         int Count { get; }
+        void Clear();
+        void Add(IIonValue item);
+        bool Remove(IIonValue item);
+        bool Contains(IIonValue item);
+        void CopyTo(IIonValue[] array, int arrayIndex);
+        IEnumerator<IIonValue> GetEnumerator();
+        IIonContainer Container { get; set; }
```
Comment (byronlin13) [inline IonDotnet/Tree/IIonContainer.cs]: We should group properties together

File: IonDotnet/Tree/IIonSequence.cs [hunk 2]
```diff
@@ -3,6 +3,6 @@
     public interface IIonSequence : IIonContainer
     {
         void RemoveAt(int index);
-        void Clear();
```
Comment (byronlin-bq) [inline IonDotnet/Tree/IIonSequence.cs]: Why is this removed?

File: IonDotnet/Tree/IIonValue.cs [hunk 3]
```diff
@@ -1,11 +1,13 @@
 ﻿using System.Collections.Generic;
-using IonDotnet.Tree.Impl;
 
 namespace IonDotnet.Tree
 {
-    public interface IIonValue
+    public interface IIonValue : IIonNull, IIonBool, IIonInt, IIonFloat,
```
Comment (byronlin13) [inline IonDotnet/Tree/IIonValue.cs:5]: Should we put the inherited interfaces in alphabetical order?

File: IonDotnet/Tree/IIonValue.cs [hunk 4]
```diff
@@ -14,9 +16,9 @@ public interface IIonValue
         string ToPrettyString();
         void WriteTo(IIonWriter writer);
         bool IsEquivalentTo(IIonValue value);
-        IonType Type { get; }
-        bool IsReadOnly { get; }
         bool IsNull { get; }
+        bool IsReadOnly { get; }
```
Comment (byronlin13) [inline IonDotnet/Tree/IIonValue.cs]: Lets group properties together

File: IonDotnet/Tree/Impl/IonValue.cs [hunk 5]
```diff
@@ -345,12 +347,83 @@ public void WriteTo(IIonWriter writer)
...
         public bool IsReadOnly => LockedFlagOn();
 
         public void MakeReadOnly() => LockedFlagOn(true);
-
-        /// <summary>The <see cref="IonType"/> of this value.</summary>
-        public abstract IonType Type { get; }
+
+        // Applicable to IonDecimal
+        public virtual decimal DecimalValue {
+            get => throw new InvalidOperationException(GetErrorMessage());
```
Comment (byronlin13) [inline IonDotnet/Tree/Impl/IonValue.cs:362]: bracket formatting

File: IonDotnet/Tree/Impl/IonValue.cs [hunk 6]
```diff
@@ -345,12 +347,83 @@ public void WriteTo(IIonWriter writer)
...
+
+        // Applicable to IonText
+        public virtual string StringValue
+        {
+            get => throw new InvalidOperationException(GetErrorMessage());
+            set => throw new InvalidOperationException(GetErrorMessage());
+        }
+
+        // Applicable to IonInt
+        public virtual IntegerSize IntegerSize => throw new InvalidOperationException(GetErrorMessage());
```
Comment (byronlin13) [inline IonDotnet/Tree/Impl/IonValue.cs:382]: Add line spacing

File: IonDotnet/Tree/IIonValue.cs (hunk 3 above)
Comment (simonz-bq) [reply IonDotnet/Tree/IIonValue.cs:5]: I think it's fine as is, since it's in the BinaryTypeId of the IonTypes

**Comment by R-maan** [comment]
Updated the fork from original
